FEATURE: Search

unclecheese opened the pull request:
Resolves: https://github.com/silverstripeltd/open-sourcerers/issues/86

## Notes

* Per ACs, spent 4 hours on an accessible autosuggest feature. It's pretty good, but may not be perfect. To turn it off, we can just comment out `search()` in the main JS file

* Reminder that this only works on the docker container. Will not work on SCP testing environment

* Will require a `docker-compose up` to run the indexing.

## Dependencies
- [ ] https://github.com/silverstripe/bambusa-theme/pull/27

Commit:
Initial integration of fulltextsearch module

Commit:
Expose mysql

Commit:
Composer update

Commit:
Add search form

Commit:
Composer update to new theme

Commit:
All ACs met

Commit:
composer update

Commit:
Linting

Commit:
Remove space sanitisation

chillu commented on app/src/Controllers/SearchController.php:
```diff
@@ -0,0 +1,115 @@
+
+        $link = $data->Link . '?' . $query;
+
+        $data->Link = $link;
```
Can you check if this is XSS?

chillu replied:
Checked a very simple one, seems fine

![image](https://user-images.githubusercontent.com/111025/62440721-757b1c00-b7a5-11e9-88ee-8b7a379c08e2.png)

chillu commented on app/src/PageController.php:
```diff
@@ -22,6 +28,37 @@ protected function init()
+        /**
+         * @return SearchForm
+         */
+        public function SearchForm(): SearchForm
```
💟

chillu commented on composer.json:
```diff
@@ -37,7 +37,9 @@
         "silverstripe/theme-fontpicker": "^1",
-        "tractorcow/silverstripe-fluent": "^4.2"
+        "tractorcow/silverstripe-fluent": "^4.2",
+        "symfony/process": "^3.2",
```
What are you using `symfony/process` for?

unclecheese replied:
* fulltextsearch needs symfony/process ^3
* framework needs symfony/process ^whatever
* ^whatever installs 4
* fulltextsearch cannot be installed
* explicitly use previous version
* no one complains

chillu reviewed (changes requested):
Lookin' good! Great job on the `search.php` workaround. A bit too much hardcoding to be general purpose, but for a sales demo it shows off what we want: Fast search.

![image](https://user-images.githubusercontent.com/111025/62337550-36409700-b529-11e9-8914-234d47aeefda.png)

Some styling issues.

Was there a reason you didn't re-enable Solr in `docker/scripts/configure-silverstripe.sh`? 

Reindex actually takes quite long on the production database locally. That's probably exacerbated by the local Docker setup being shit slow, but still not great since it delays demo creation. One of the reasons is that we're indexing everything *twice* through the built in ElementalSolrIndex. Just looking at how we can disable that thing.

chillu committed:
Re-enable automatic index

chillu committed:
Whitelist search index

Avoid double indexing, see https://github.com/dnadesign/silverstripe-elemental/issues/709

chillu committed:
Don't index draft content

It's a security issue waiting to happen,
and doubles our indexing time (and time-to-demo-ready)

chillu commented:
I've pushed some commits which disable the double search index. Done some upstream PRs to https://github.com/dnadesign/silverstripe-elemental as well

Commit:
New getElementsForSearch method

chillu committed:
Fixed linter

Commit:
Fix encoding with HTMLText casting, newline sanitisation

chillu commented:
The search results still have encoding errors, but it's good enough for now

![image](https://user-images.githubusercontent.com/111025/62440677-49f83180-b7a5-11e9-9f5c-0cb4bb303dc4.png)

unclecheese commented:
Need to reindex to resolve the encoding issue. It's an issue with the write, not the read. :)
